### ISSUE TITLE
Improve error logging and reporting

### ISSUE THREAD

**Issue by mwest1066** [PrairieLearn/PrairieLearn#592]
Use `RAISE EXCEPTION` inside PL/pgSQL code to return user-visible errors.

**Comment by mwest1066** [comment]
Comment from an error thread on #613:

I'm thinking about the various ways that something should go wrong, how those errors should be reported, and who they should be reported to. It seems to me we have three different types of errors: (1) system errors, like a missing `grading_id` or a failed write to an S3 bucket, (2) autograder errors, like invalid JSON output by the autograder, and (3) errors in student code, like a missing file. Similarly, we have three different types of people to report to: (A) system admins, (B) course instructors, (C) the student. I think that all error types need to be reported to C, types 1 and 2 need to be reported to B, and only type 1 errors should be reported to A. However, aggregate statistics on type 2 errors might be wanted by A to see which courses are having issues. Initially it's more important that we record error data, and we can worry about reporting interfaces later.

It's possible in the error classification scheme above that type 3 errors are not really "errors", but are just handled by the autograder in the regular way with feedback. I'm not sure about this, however. For example, if there is a missing file in the student submission, is this a different kind of error than a compile error, or a failing testcase? Maybe we want to just handle all of them the same.

When we are reporting errors, it's important to be clear about who's "fault" the error is. Maybe we can report these as "system error", "autograder error", etc. We also shouldn't leak private information, so course instructors shouldn't see details of our AWS infrastructure, and students shouldn't see autograder details.

One useful technique that PL uses for errors already is to generate a random unique errorId (see https://github.com/PrairieLearn/PrairieLearn/blob/master/pages/error/error.js#L14). This is reported to the student without additional details, but then the log files contain the same errorId with full debug output (stack traces, internal variable states, etc). This way the sysadmins can get a bug report saying "I did XXX and it said errorId: 4K8XU81", and then it's easy to go look up the detailed information in the logs.

I'm thinking that when we have an error, we want to:

1. Generate a random errorId. We could actually just make these UUIDs.

2. Set the errorType to either "system", "course", or "student" (maybe we only just use the first two?).

3. Set systemErrorMessage, systemErrorData, courseErrorMessage, courseErrorData, studentErrorMessage, studentErrorData. These might all be the same, or they might contain variable levels of detail.

4. Record all of the above along with an "error: true" flag in the job output data.

Example 1: results JSON doesn't parse:

* `errorType = 'course'`
* `systemErrorMessage = courseErrorMessage = 'Unable to parse result JSON'`
* `systemErrorData = courseErrorData = {rawData: data, parseError: msg}'
* `studentErrorMessage = 'Autograder error'`
* `studentErrorData = null`

Example 2: unable to write to S3:

* `errorType = 'system'`
* `systemErrorMessage = 'Error writing to S3 bucket'`
* `systemErrorData = {bucket: 'XXXX', errorMsg: msg}`
* `courseErrorMessage = studentErrorMessage = 'System error'`
* `courseErrorData = studentErrorData = null`

In all cases we report the errorUuid to everyone (course staff, students, etc), and we only report the appropriate error message and error data.

To support this error reporting infrastructure we should eventually make a new DB table called `errors` to track all errors. We don't need to do this as part of the autograder branch, however. This will also address #592.

**Comment by mwest1066** [comment]
Here's an updated concept. As @nwalters512 suggested in a slack thread, we won't handle student input errors via this error mechanism. They will get handled in the same way as any other student errors on questions, such as incorrect answers.

We thus have two levels of errors: system errors and course errors. Course errors are those caused by bad course code in some way, and need to be fixed by course staff (e.g., course question code throws an exception when trying to grade a question). System errors are caused by PL and course staff can't fix them (e.g., error writing to S3 bucket).

For each error that occurs we insert a new row into the `errors` table below. We fill in as may of the `course_id`, `user_id`, etc, as are relevant for this error. If the error was associated with a particular grading job then we write the `error_id` into the corresponding `submission` and `grading_log`.

When displaying errors to users, admin users can always see the `message` and `data`. If it's a course error then course staff can see the `message` and `data`. Students are shown the `user_message`. If it's a system error then course staff are also shown the `user_message`. Everyone is always shown the `display_id`.

```sql
CREATE TABLE errors (
    id BIGSERIAL,
    display_id TEXT, -- generated by https://github.com/PrairieLearn/PrairieLearn/blob/master/pages/error/error.js#L14
    date TIMESTAMPTZ,
    course_caused BOOLEAN, --  is a course error (`TRUE`) or system error (`FALSE`)?
    message text,
    data jsonb,
    user_message text,
    course_id BIGINT REFERENCES pl_courses ON DELETE SET NULL ON UPDATE CASCADE,
    course_instance_id BIGINT REFERENCES course_instances ON DELETE SET NULL ON UPDATE CASCADE,
    question_id BIGINT REFERENCES questions ON DELETE SET NULL ON UPDATE CASCADE,
    assessment_id BIGINT REFERENCES assessments ON DELETE SET NULL ON UPDATE CASCADE,
    authn_user_id BIGINT REFERENCES users ON DELETE SET NULL ON UPDATE CASCADE,
    user_id BIGINT REFERENCES users ON DELETE SET NULL ON UPDATE CASCADE,
);
```

Add a column to `submissions` and `grading_logs` like:

```sql
ADD COLUMN error_id BIGINT REFERENCES errors ON DELETE SET NULL ON UPDATE CASCADE;
```

**Comment by mwest1066** [comment]
Implemented in 7086c365fee190dd805391a8356b8e0fee691099
